Requests is a Python Library that allows you to send <code>HTTP/1.1</code> requests easily. We can import the library as follows:


In [59]:
pip install requests

Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.



In [60]:
import requests

We will also use the following libraries:


In [61]:
import os 
from PIL import Image
from IPython.display import IFrame

You can make a <code>GET</code> request via the method <code>get</code> to [www.ibm.com](http://www.ibm.com/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01):


In [62]:

url='https://www.ibm.com/'
r=requests.get(url)

print( r.status_code ) # 100 - działa, jeszcze nie wiadomo; 200 - success; 300 przekierowanie; 400 brak
print(r.request.headers)
print("request body:", r.request.body )


200
{'User-Agent': 'python-requests/2.31.0', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Cookie': '_abck=64AEC157D3F6BCFB5E26FEDB2FD8E460~-1~YAAQd+IlF2YpPwSKAQAAXgmXDgqniJFpdATpOHulfsNQ4gPUEo9fPyPuzbFgrbrtD0ftdE9+4ZAu+mGUynpBq8rRrVqbK/LLrW/4yhHtPo+91zfVq8Frvw1hs0qqW89/Yd6GTOrYGRfPP5XQIKv0v+Z4BIzXuNYbei92MV0RxNdXW9QK2GHUeRryJJ0NiEj8AdBzxCDhrwJ2xPblcdY3NPCUc3dd3CDxY+fPY7z1ytUZIwdozLNqjVgwVP8gwurXo6+Z3wg8A78Fpe0PNjWkixlnrmMSUHXMOczIdv0dIyeMuFB0waEUDAZjGOnxTJjIkGiTY6kr7nlH3eEtqVwa4bbulKLENSqB4Pe0rJlAdAi1lDVNuxc=~-1~-1~-1; bm_sz=47A8A93052DC251EC535F7E48BDACEDB~YAAQd+IlF2cpPwSKAQAAXwmXDhS1VZq+XFQ67p77baVFXI62/xLSXhQnUntZjsxZhF408iUxNYGTH5+lqtv7tUvls2fNGGd2nuHG1XM+uWjYOBgNULv47wljpBPHwz4afEDJ1D3TP34MmJVpa++P/7EWW2GU09QSv0sES8tGA7xTOykd7LcSiseNh2KoG/4hYKoAqPGi8rZoFI6lMDQ5fLN3YKsTKw499NW9fuO41HNThgf2fOLL1QYbQn5xlijlNSfeNdCZ7SSOP1ct0+3/ZFJ3/6Q4OTKEoxWbhrJuGnQ=~3425585~4534850'}
request body: None


In [63]:
# Use single quotation marks for defining string
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png'

r = requests.get(url)
print( r.status_code, r.headers['Content-Type'] )
with open(path,'wb') as f:
    f.write(r.content)
    


200 image/png


<h3>Question 1: write <a href="https://www.gnu.org/software/wget/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01"><code> wget </code></a></h3>


In the previous section, we used the <code>wget</code> function to retrieve content from the web server as shown below.  Write the python code to perform the same task. The code should be the same as the one used to download the image, but the file name should be <code>'Example1.txt'</code>.


<code>!wget -O /resources/data/Example1.txt <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt></code>


<details><summary>Click here for the solution</summary>

```python
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
path=os.path.join(os.getcwd(),'example1.txt')
r=requests.get(url)
with open(path,'wb') as f:
    f.write(r.content)

```

</details>


In [66]:
r=requests.get(url_get,params=payload)

We can print out the <code>URL</code> and see the name and values.


In [67]:
r.url

'http://httpbin.org/get?name=Joseph&ID=123'

There is no request body.


In [68]:
print("request body:", r.request.body)

request body: None


We can print out the status code.


In [69]:
print(r.status_code)

200


We can view the response as text:


In [70]:
print(r.text)

{
  "args": {
    "ID": "123", 
    "name": "Joseph"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Host": "httpbin.org", 
    "User-Agent": "python-requests/2.31.0", 
    "X-Amzn-Trace-Id": "Root=1-64e0eb48-0ab05a163430218e26f23f31"
  }, 
  "origin": "37.30.2.15", 
  "url": "http://httpbin.org/get?name=Joseph&ID=123"
}



We can look at the <code>'Content-Type'</code>.


In [71]:
r.headers['Content-Type']

'application/json'

As the content <code>'Content-Type'</code> is in the <code>JSON</code> format we can use the method <code>json()</code>, it returns a Python <code>dict</code>:


In [72]:
r.json()

{'args': {'ID': '123', 'name': 'Joseph'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.31.0',
  'X-Amzn-Trace-Id': 'Root=1-64e0eb48-0ab05a163430218e26f23f31'},
 'origin': '37.30.2.15',
 'url': 'http://httpbin.org/get?name=Joseph&ID=123'}

The key <code>args</code> has the name and values:


In [73]:
r.json()['args']

{'ID': '123', 'name': 'Joseph'}

<h2 id="POST">Post Requests  </h2>


Like a <code>GET</code> request, a <code>POST</code> is used to send data to a server, but the <code>POST</code> request sends the data in a request body. In order to send the Post Request in Python, in the <code>URL</code> we change the route to <code>POST</code>:


In [74]:
url_post='http://httpbin.org/post'

This endpoint will expect data as a file or as a form. A form is convenient way to configure an HTTP request to send data to a server.


To make a <code>POST</code> request we use the <code>post()</code> function, the variable <code>payload</code> is passed to the parameter <code> data </code>:


In [75]:
r_post=requests.post(url_post,data=payload)

Comparing the URL from the response object of the <code>GET</code> and <code>POST</code> request we see the <code>POST</code> request has no name or value pairs.


In [76]:
print("POST request URL:",r_post.url )
print("GET request URL:",r.url)

POST request URL: http://httpbin.org/post
GET request URL: http://httpbin.org/get?name=Joseph&ID=123


We can compare the <code>POST</code> and <code>GET</code> request body, we see only the <code>POST</code> request has a body:


In [77]:
print("POST request body:",r_post.request.body)
print("GET request body:",r.request.body)

POST request body: name=Joseph&ID=123
GET request body: None


We can view the form as well:


In [78]:
r_post.json()['form']

{'ID': '123', 'name': 'Joseph'}

There is a lot more you can do. Check out <a href="https://requests.readthedocs.io/en/master/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Requests </a> for more.


<hr>


## Authors

<p><a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Joseph Santarcangelo</a> <br>A Data Scientist at IBM, and holds a PhD in Electrical Engineering. His research focused on using Machine Learning, Signal Processing, and Computer Vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.</p>


### Other Contributors

<a href="https://www.linkedin.com/in/jiahui-mavis-zhou-a4537814a?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Mavis Zhou</a>


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description           |
| ----------------- | ------- | ---------- | ---------------------------- |
| 2023-06-07        | 2.3     |Akansha Yadav| Spell Check                 |
| 2021-12-20        | 2.1     | Malika     | Updated the links            |
| 2020-09-02        | 2.0     | Simran     | Template updates to the file |
|                   |         |            |                              |
|                   |         |            |                              |

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
